File: mts/src/mts/storage/sqlite_store.py
from __future__ import annotations

import sqlite3
from pathlib import Path
from typing import Any


class SQLiteStore:
    def __init__(self, db_path: Path):
        self.db_path = db_path
        self.db_path.parent.mkdir(parents=True, exist_ok=True)

    def connect(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self.db_path)
        conn.row_factory = sqlite3.Row
        conn.execute("PRAGMA foreign_keys=ON;")
        return conn

    def migrate(self, migrations_dir: Path) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS schema_migrations (
                    version TEXT PRIMARY KEY,
                    applied_at TEXT NOT NULL DEFAULT (datetime('now'))
                );
                """
            )
            for migration in sorted(migrations_dir.glob("*.sql")):
                applied = conn.execute(
                    "SELECT version FROM schema_migrations WHERE version = ?",
                    (migration.name,),
                ).fetchone()
                if applied:
                    continue
                conn.executescript(migration.read_text(encoding="utf-8"))
                conn.execute("INSERT INTO schema_migrations(version) VALUES (?)", (migration.name,))

    def create_run(self, run_id: str, scenario: str, generations: int, executor_mode: str, agent_provider: str = "") -> None:
        with self.connect() as conn:
            conn.execute(
                """
                INSERT OR IGNORE INTO runs(run_id, scenario, target_generations, executor_mode, status, agent_provider)
                VALUES (?, ?, ?, ?, 'running', ?)
                """,
                (run_id, scenario, generations, executor_mode, agent_provider),
            )

    def generation_exists(self, run_id: str, generation_index: int) -> bool:
        with self.connect() as conn:
            row = conn.execute(
                "SELECT 1 FROM generations WHERE run_id = ? AND generation_index = ?",
                (run_id, generation_index),
            ).fetchone()
            return row is not None

    def upsert_generation(
        self,
        run_id: str,
        generation_index: int,
        mean_score: float,
        best_score: float,
        elo: float,
        wins: int,
        losses: int,
        gate_decision: str,
        status: str,
    ) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                INSERT INTO generations(
                    run_id, generation_index, mean_score, best_score, elo, wins, losses, gate_decision, status
                )
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(run_id, generation_index) DO UPDATE SET
                    mean_score = excluded.mean_score,
                    best_score = excluded.best_score,
                    elo = excluded.elo,
                    wins = excluded.wins,
                    losses = excluded.losses,
                    gate_decision = excluded.gate_decision,
                    status = excluded.status,
                    updated_at = datetime('now')
                """,
                (run_id, generation_index, mean_score, best_score, elo, wins, losses, gate_decision, status),
            )

    def insert_match(
        self,
        run_id: str,
        generation_index: int,
        seed: int,
        score: float,
        passed_validation: bool,
        validation_errors: str,
    ) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                INSERT INTO matches(run_id, generation_index, seed, score, passed_validation, validation_errors)
                VALUES (?, ?, ?, ?, ?, ?)
                """,
                (run_id, generation_index, seed, score, int(passed_validation), validation_errors),
            )

    def append_agent_output(self, run_id: str, generation_index: int, role: str, content: str) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                INSERT INTO agent_outputs(run_id, generation_index, role, content)
                VALUES (?, ?, ?, ?)
                """,
                (run_id, generation_index, role, content),
            )

    def get_agent_outputs_by_role(self, run_id: str, role: str) -> list[dict[str, object]]:
        """Return agent_outputs rows for a given run and role, ordered by generation."""
        with self.connect() as conn:
            rows = conn.execute(
                """
                SELECT generation_index, role, content
                FROM agent_outputs
                WHERE run_id = ? AND role = ?
                ORDER BY generation_index
                """,
                (run_id, role),
            ).fetchall()
            return [dict(r) for r in rows]

    def append_agent_role_metric(
        self,
        run_id: str,
        generation_index: int,
        role: str,
        model: str,
        input_tokens: int,
        output_tokens: int,
        latency_ms: int,
        subagent_id: str,
        status: str,
    ) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                INSERT INTO agent_role_metrics(
                    run_id, generation_index, role, model, input_tokens, output_tokens, latency_ms, subagent_id, status
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    run_id,
                    generation_index,
                    role,
                    model,
                    input_tokens,
                    output_tokens,
                    latency_ms,
                    subagent_id,
                    status,
                ),
            )

    def append_recovery_marker(
        self,
        run_id: str,
        generation_index: int,
        decision: str,
        reason: str,
        retry_count: int,
    ) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                INSERT INTO generation_recovery(run_id, generation_index, decision, reason, retry_count)
                VALUES (?, ?, ?, ?, ?)
                """,
                (run_id, generation_index, decision, reason, retry_count),
            )

    def get_matches_for_run(self, run_id: str) -> list[dict[str, Any]]:
        """Return all match records for a run, ordered by generation and seed."""
        with self.connect() as conn:
            rows = conn.execute(
                "SELECT * FROM matches WHERE run_id = ? ORDER BY generation_index, seed",
                (run_id,),
            ).fetchall()
            return [dict(row) for row in rows]

    def get_generation_metrics(self, run_id: str) -> list[dict[str, Any]]:
        """Return all generation records for a run, ordered by generation index."""
        with self.connect() as conn:
            rows = conn.execute(
                "SELECT * FROM generations WHERE run_id = ? ORDER BY generation_index",
                (run_id,),
            ).fetchall()
            return [dict(row) for row in rows]

    def get_generation_trajectory(self, run_id: str) -> list[dict[str, Any]]:
        """Return generation trajectory with score deltas."""
        with self.connect() as conn:
            rows = conn.execute(
                """
                SELECT generation_index, mean_score, best_score, elo, gate_decision
                FROM generations
                WHERE run_id = ? AND status = 'completed'
                ORDER BY generation_index
                """,
                (run_id,),
            ).fetchall()
            result = []
            prev_best = 0.0
            for row in rows:
                d = dict(row)
                d["delta"] = round(d["best_score"] - prev_best, 6)
                prev_best = d["best_score"]
                result.append(d)
            return result

    def get_strategy_score_history(self, run_id: str) -> list[dict[str, Any]]:
        """Return strategy content with scores, joining agent_outputs and generations."""
        with self.connect() as conn:
            rows = conn.execute(
                """
                SELECT ao.generation_index, ao.content, g.best_score, g.gate_decision
                FROM agent_outputs ao
                JOIN generations g ON ao.run_id = g.run_id AND ao.generation_index = g.generation_index
                WHERE ao.run_id = ? AND ao.role = 'competitor' AND g.status = 'completed'
                ORDER BY ao.generation_index
                """,
                (run_id,),
            ).fetchall()
            return [dict(row) for row in rows]

    def save_knowledge_snapshot(
        self,
        scenario: str,
        run_id: str,
        best_score: float,
        best_elo: float,
        playbook_hash: str,
        agent_provider: str = "",
        rlm_enabled: bool = False,
    ) -> None:
        with self.connect() as conn:
            conn.execute(
                """
                INSERT INTO knowledge_snapshots(
                    scenario, run_id, best_score, best_elo, playbook_hash, agent_provider, rlm_enabled
                ) VALUES (?, ?, ?, ?, ?, ?, ?)
                """,
                (scenario, run_id, best_score, best_elo, playbook_hash, agent_provider, int(rlm_enabled)),
            )

    def get_best_knowledge_snapshot(self, scenario: str) -> dict[str, Any] | None:
        with self.connect() as conn:
            row = conn.execute(
                """
                SELECT scenario, run_id, best_score, best_elo, playbook_hash, created_at
                FROM knowledge_snapshots
                WHERE scenario = ?
                ORDER BY best_score DESC
                LIMIT 1
                """,
                (scenario,),
            ).fetchone()
            return dict(row) if row else None

    def get_ecosystem_snapshots(self, scenario: str) -> list[dict[str, Any]]:
        """Return all knowledge snapshots for a scenario with provider info, ordered by created_at ASC."""
        with self.connect() as conn:
            rows = conn.execute(
                """
                SELECT scenario, run_id, best_score, best_elo, playbook_hash, agent_provider, rlm_enabled, created_at
                FROM knowledge_snapshots
                WHERE scenario = ?
                ORDER BY id ASC
                """,
                (scenario,),
            ).fetchall()
            return [dict(row) for row in rows]

    def get_best_competitor_output(self, scenario: str) -> str | None:
        """Return the competitor output from the best-scoring generation across all runs for a scenario."""
        with self.connect() as conn:
            row = conn.execute(
                """
                SELECT ao.content
                FROM agent_outputs ao
                JOIN generations g ON ao.run_id = g.run_id AND ao.generation_index = g.generation_index
                JOIN runs r ON g.run_id = r.run_id
                WHERE r.scenario = ? AND ao.role = 'competitor' AND g.status = 'completed'
                ORDER BY g.best_score DESC
                LIMIT 1
                """,
                (scenario,),
            ).fetchone()
            return row["content"] if row else None

    def count_completed_runs(self, scenario: str) -> int:
        """Return count of completed runs for a scenario."""
        with self.connect() as conn:
            row = conn.execute(
                "SELECT COUNT(*) as cnt FROM runs WHERE scenario = ? AND status = 'completed'",
                (scenario,),
            ).fetchone()
            return row["cnt"] if row else 0

    def mark_run_completed(self, run_id: str) -> None:
        with self.connect() as conn:
            conn.execute("UPDATE runs SET status = 'completed', updated_at = datetime('now') WHERE run_id = ?", (run_id,))

    # -- Human feedback --

    def insert_human_feedback(
        self,
        scenario_name: str,
        agent_output: str,
        human_score: float | None = None,
        human_notes: str = "",
        generation_id: str | None = None,
    ) -> int:
        """Store human feedback on an agent task output. Returns the row id."""
        if human_score is not None and not (0.0 <= human_score <= 1.0):
            raise ValueError(f"human_score must be in [0.0, 1.0], got {human_score}")
        with self.connect() as conn:
            cursor = conn.execute(
                """
                INSERT INTO human_feedback(scenario_name, generation_id, agent_output, human_score, human_notes)
                VALUES (?, ?, ?, ?, ?)
                """,
                (scenario_name, generation_id, agent_output, human_score, human_notes),
            )
            return cursor.lastrowid or 0

    def get_human_feedback(self, scenario_name: str, limit: int = 10) -> list[dict[str, Any]]:
        """Retrieve recent human feedback for a scenario."""
        with self.connect() as conn:
            rows = conn.execute(
                """
                SELECT id, scenario_name, generation_id, agent_output, human_score, human_notes, created_at
                FROM human_feedback
                WHERE scenario_name = ?
                ORDER BY created_at DESC
                LIMIT ?
                """,
                (scenario_name, limit),
            ).fetchall()
            return [dict(r) for r in rows]

    def get_calibration_examples(self, scenario_name: str, limit: int = 5) -> list[dict[str, Any]]:
        """Retrieve feedback with both score and notes — suitable for judge calibration."""
        with self.connect() as conn:
            rows = conn.execute(
                """
                SELECT id, scenario_name, agent_output, human_score, human_notes, created_at
                FROM human_feedback
                WHERE scenario_name = ? AND human_score IS NOT NULL AND human_notes != ''
                ORDER BY created_at DESC
                LIMIT ?
                """,
                (scenario_name, limit),
            ).fetchall()
            return [dict(r) for r in rows]

    # ---- Task Queue CRUD ----

    def enqueue_task(
        self,
        task_id: str,
        spec_name: str,
        priority: int = 0,
        config: dict[str, Any] | None = None,
        scheduled_at: str | None = None,
    ) -> None:
        """Add a task to the queue."""
        import json as _json

        config_json = _json.dumps(config) if config else None
        with self.connect() as conn:
            conn.execute(
                """
                INSERT INTO task_queue(id, spec_name, priority, config_json, scheduled_at)
                VALUES (?, ?, ?, ?, ?)
                """,
                (task_id, spec_name, priority, config_json, scheduled_at),
            )

    def dequeue_task(self) -> dict[str, Any] | None:
        """Claim the highest-priority pending task.

        Returns the task row as a dict, or None if queue is empty.
        Uses a single UPDATE with subquery for true atomic dequeue —
        prevents double-processing under concurrent access.
        """
        with self.connect() as conn:
            # Atomic: SELECT the best candidate, then UPDATE in one transaction.
            # SQLite's write lock on the transaction prevents two connections
            # from claiming the same row.
            row = conn.execute(
                """
                SELECT id FROM task_queue
                WHERE status = 'pending'
                  AND (scheduled_at IS NULL OR scheduled_at <= datetime('now'))
                ORDER BY priority DESC, created_at ASC
                LIMIT 1
                """,
            ).fetchone()
            if not row:
                return None

            task_id = row["id"]
            conn.execute(
                """
                UPDATE task_queue
                SET status = 'running',
                    started_at = datetime('now'),
                    updated_at = datetime('now')
                WHERE id = ? AND status = 'pending'
                """,
                (task_id,),
            )
            # Check we actually claimed it (another runner might have beaten us)
            if conn.execute("SELECT changes()").fetchone()[0] == 0:
                return None

            updated = conn.execute(
                "SELECT * FROM task_queue WHERE id = ?", (task_id,)
            ).fetchone()
            return dict(updated) if updated else None

    def complete_task(
        self,
        task_id: str,
        best_score: float,
        best_output: str,
        total_rounds: int,
        met_threshold: bool,
        result_json: str | None = None,
    ) -> None:
        """Mark a task as completed with results."""
        with self.connect() as conn:
            conn.execute(
                """
                UPDATE task_queue
                SET status = 'completed',
                    completed_at = datetime('now'),
                    updated_at = datetime('now'),
                    best_score = ?,
                    best_output = ?,
                    total_rounds = ?,
                    met_threshold = ?,
                    result_json = ?
                WHERE id = ?
                """,
                (best_score, best_output, total_rounds, 1 if met_threshold else 0, result_json, task_id),
            )

    def fail_task(self, task_id: str, error: str) -> None:
        """Mark a task as failed."""
        with self.connect() as conn:
            conn.execute(
                """
                UPDATE task_queue
                SET status = 'failed',
                    completed_at = datetime('now'),
                    updated_at = datetime('now'),
                    error = ?
                WHERE id = ?
                """,
                (error, task_id),
            )

    def get_task(self, task_id: str) -> dict[str, Any] | None:
        """Get a task by ID."""
        with self.connect() as conn:
            row = conn.execute(
                "SELECT * FROM task_queue WHERE id = ?", (task_id,)
            ).fetchone()
            return dict(row) if row else None

    def list_tasks(
        self,
        status: str | None = None,
        spec_name: str | None = None,
        limit: int = 50,
    ) -> list[dict[str, Any]]:
        """List tasks with optional filters."""
        query = "SELECT * FROM task_queue WHERE 1=1"
        params: list[Any] = []
        if status:
            query += " AND status = ?"
            params.append(status)
        if spec_name:
            query += " AND spec_name = ?"
            params.append(spec_name)
        query += " ORDER BY created_at DESC LIMIT ?"
        params.append(limit)
        with self.connect() as conn:
            rows = conn.execute(query, params).fetchall()
            return [dict(r) for r in rows]

    def pending_task_count(self) -> int:
        """Count pending tasks in the queue."""
        with self.connect() as conn:
            row = conn.execute(
                "SELECT COUNT(*) as cnt FROM task_queue WHERE status = 'pending'"
            ).fetchone()
            return row["cnt"] if row else 0

File: ts/src/storage/sqlite-store.ts
import Database from "better-sqlite3";

import type {
  AgentOutputRow,
  GenerationRow,
  HumanFeedbackRow,
  MatchRow,
  RecordMatchOpts,
  RunRow,
  TaskQueueRow,
  TrajectoryRow,
  UpsertGenerationOpts,
} from "./storage-contracts.js";
import {
  completeStoreTask,
  countPendingStoreTasks,
  dequeueStoreTask,
  enqueueStoreTask,
  failStoreTask,
  getStoreTask,
} from "./storage-task-queue-facade.js";
import {
  appendStoreAgentOutput,
  countStoreCompletedRuns,
  createStoreRun,
  getStoreAgentOutputs,
  getStoreBestGenerationForScenario,
  getStoreBestMatchForScenario,
  getStoreGenerations,
  getStoreMatchesForGeneration,
  getStoreMatchesForRun,
  getStoreRun,
  getStoreScoreTrajectory,
  listStoreRuns,
  listStoreRunsForScenario,
  recordStoreMatch,
  upsertStoreGeneration,
  updateStoreRunStatus,
} from "./storage-generation-run-facade.js";
import {
  getStoreCalibrationExamples,
  getStoreHumanFeedback,
  insertStoreHumanFeedback,
} from "./storage-human-feedback-facade.js";
import { migrateDatabase } from "./storage-migration-workflow.js";

export function configureSqliteDatabase(db: Pick<Database.Database, "pragma">): void {
  db.pragma("journal_mode = WAL");
  db.pragma("foreign_keys = ON");
}

export class SQLiteStore {
  #db: Database.Database;

  constructor(dbPath: string) {
    this.#db = new Database(dbPath);
    configureSqliteDatabase(this.#db);
  }

  migrate(migrationsDir: string): void {
    migrateDatabase(this.#db, migrationsDir);
  }

  enqueueTask(
    id: string,
    specName: string,
    priority = 0,
    config?: Record<string, unknown>,
    scheduledAt?: string,
  ): void {
    enqueueStoreTask(this.#db, id, specName, priority, config, scheduledAt);
  }

  dequeueTask(): TaskQueueRow | null {
    return dequeueStoreTask(this.#db);
  }

  completeTask(
    taskId: string,
    bestScore: number,
    bestOutput: string,
    totalRounds: number,
    metThreshold: boolean,
    resultJson?: string,
  ): void {
    completeStoreTask(
      this.#db,
      taskId,
      bestScore,
      bestOutput,
      totalRounds,
      metThreshold,
      resultJson,
    );
  }

  failTask(taskId: string, error: string): void {
    failStoreTask(this.#db, taskId, error);
  }

  pendingTaskCount(): number {
    return countPendingStoreTasks(this.#db);
  }

  getTask(taskId: string): TaskQueueRow | null {
    return getStoreTask(this.#db, taskId);
  }

  insertHumanFeedback(
    scenarioName: string,
    agentOutput: string,
    humanScore?: number | null,
    humanNotes = "",
    generationId?: string | null,
  ): number {
    return insertStoreHumanFeedback(
      this.#db,
      scenarioName,
      agentOutput,
      humanScore,
      humanNotes,
      generationId,
    );
  }

  getHumanFeedback(scenarioName: string, limit = 10): HumanFeedbackRow[] {
    return getStoreHumanFeedback(this.#db, scenarioName, limit);
  }

  getCalibrationExamples(scenarioName: string, limit = 5): HumanFeedbackRow[] {
    return getStoreCalibrationExamples(this.#db, scenarioName, limit);
  }

  createRun(
    runId: string,
    scenario: string,
    generations: number,
    executorMode: string,
    agentProvider = "",
  ): void {
    createStoreRun(this.#db, runId, scenario, generations, executorMode, agentProvider);
  }

  getRun(runId: string): RunRow | null {
    return getStoreRun(this.#db, runId);
  }

  updateRunStatus(runId: string, status: string): void {
    updateStoreRunStatus(this.#db, runId, status);
  }

  upsertGeneration(runId: string, generationIndex: number, opts: UpsertGenerationOpts): void {
    upsertStoreGeneration(this.#db, runId, generationIndex, opts);
  }

  getGenerations(runId: string): GenerationRow[] {
    return getStoreGenerations(this.#db, runId);
  }

  countCompletedRuns(scenario: string): number {
    return countStoreCompletedRuns(this.#db, scenario);
  }

  getBestGenerationForScenario(scenario: string): (GenerationRow & { run_id: string }) | null {
    return getStoreBestGenerationForScenario(this.#db, scenario);
  }

  getBestMatchForScenario(scenario: string): MatchRow | null {
    return getStoreBestMatchForScenario(this.#db, scenario);
  }

  recordMatch(runId: string, generationIndex: number, opts: RecordMatchOpts): void {
    recordStoreMatch(this.#db, runId, generationIndex, opts);
  }

  getMatchesForRun(runId: string): MatchRow[] {
    return getStoreMatchesForRun(this.#db, runId);
  }

  appendAgentOutput(runId: string, generationIndex: number, role: string, content: string): void {
    appendStoreAgentOutput(this.#db, runId, generationIndex, role, content);
  }

  getAgentOutputs(runId: string, generationIndex: number): AgentOutputRow[] {
    return getStoreAgentOutputs(this.#db, runId, generationIndex);
  }

  getScoreTrajectory(runId: string): TrajectoryRow[] {
    return getStoreScoreTrajectory(this.#db, runId);
  }

  listRuns(limit = 50, scenario?: string): RunRow[] {
    return listStoreRuns(this.#db, limit, scenario);
  }

  listRunsForScenario(scenario: string): RunRow[] {
    return listStoreRunsForScenario(this.#db, scenario);
  }

  getMatchesForGeneration(runId: string, generationIndex: number): MatchRow[] {
    return getStoreMatchesForGeneration(this.#db, runId, generationIndex);
  }

  close(): void {
    this.#db.close();
  }
}
